controlplane/internal/controller/agentcontrolplane_controller.go

```diff
@@ -452,3 +450,39 @@ func (r *AgentControlPlaneReconciler) generateAgentBootstrapConfig(ctx context.C
 
 	return reference.GetReference(r.Scheme, bootstrapConfig)
 }
+
+func (r *AgentControlPlaneReconciler) ensurePullSecret(ctx context.Context, acp *controlplanev1alpha1.AgentControlPlane) error {
+	if acp.Spec.AgentConfigSpec.PullSecretRef != nil {
+		return nil
+	}
+
+	secret := GenerateSecret(placeholderPullSecretName, acp.Namespace)
+	if err := controllerutil.SetOwnerReference(acp, secret, r.Scheme); err != nil {
+		return err
+	}
+
+	if err := r.Client.Create(ctx, secret); err != nil {
+		return err
+	}
+	acp.Spec.AgentConfigSpec.PullSecretRef = &corev1.LocalObjectReference{Name: secret.Name}
+	return r.Client.Update(ctx, acp)
+}
+
+// Assisted-service expects the pull secret to
+// 1. Have .dockerconfigjson as a key
+// 2. Have the value of .dockerconfigjson be a base64-encoded JSON
+// 3. The JSON must have the key "auths" followed by repository with an "auth" key
+func GenerateSecret(name, namespace string) *corev1.Secret {
```

Comment:
Does this need to be in `utils`? Is it going to be used anywhere else? If not, we can create a new file in the package where it's used, i.e. `auth.go`

Reply:
Moved here https://github.com/openshift-assisted/cluster-api-agent/pull/43